core/src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -1901,14 +1901,68 @@ sepBy
 -------------------------------------------------------------------------------
 --
 -- | See 'Streamly.Internal.Data.Parser.sequence'.
---
--- /Unimplemented/
---
 {-# INLINE sequence #-}
-sequence ::
-    -- Foldable t =>
-    Fold m b c -> t (Parser m a b) -> Parser m a c
-sequence _f _p = undefined
+sequence :: MonadThrow m => Fold m b c -> D.Stream m (Parser m a b) -> Parser m a c
+sequence (Fold fstep finitial fextract) (D.Stream sstep sstate) =
+    Parser step initial extract
+    where
+    initial = do
+        fres <- finitial
+        case fres of
+            FL.Partial fs -> return $ IPartial (Nothing', sstate, fs)
+            FL.Done c -> return $ IDone c
+
+    -- state does not contain any parser
+    -- yield a new parser from the stream
+    step (Nothing', ss, fs) _ = do
+        sres <- sstep defState ss
+        case sres of
+            D.Yield p ss' -> return $ Continue 1 (Just' $ Left' p, ss', fs)
+            D.Stop -> do
+                c <- fextract fs
+                return $ Done 1 c
+            D.Skip ss' -> return $ Continue 1 (Nothing', ss', fs)
+
+    -- state contains a parser that has not yet been initialized
+    step (Just' (Left' (Parser pstep pinit pextr)), ss, fs) _ = do
+        ps <- pinit
+        case ps of
+            IPartial ps' -> return $ Continue 1 (Just' $ Right' $ Parser pstep (return $ IPartial ps') pextr, ss, fs)
+            IDone b -> do
+                fres <- fstep fs b
+                case fres of
+                    FL.Partial fs' ->
+                        return $ Continue 1 (Nothing', ss, fs')
+                    FL.Done c -> return $ Done 1 c
+            IError err -> throwM $ ParseError err
+
+    -- state contains a parser that has been initialized
+    -- and that needs to be stepped forward through the input
+    step (Just' (Right' (Parser pstep pinit pextr)), ss, fs) a = do
+        ps <- pinit
+        case ps of
+            IPartial ps' -> do
+                pres <- pstep ps' a
+                case pres of
+                    Partial n ps'' -> return $ Continue n (Just' $ Right' $ Parser pstep (return $ IPartial ps'') pextr, ss, fs)
```

Comment:
`Partial` instead of `Continue`?

Reply:
So the general pattern I here is that when the inner Parser returns a Partial then the step should also be a Partial. This would flow directly into the extract step where, as you mentioned, I should extract the Parser state check if it is Partial, apply it to the fold and then return the result.